File: libs/interchange/src/manifest.cpp
// Faithful port of paleo_workbench/interchange/package/manifest.py.
// Key order in to_dict() is the on-disk manifest layout — do not reorder.

#include <pwb/interchange/manifest.hpp>

#include <pwb/interchange/path_safety.hpp>

#include "py_compat.hpp"

#include <cstdio>
#include <stdexcept>

namespace pwb::interchange {

namespace {
using detail::py_llong;
using detail::py_str;
using detail::repr_scalar;
}  // namespace

Json PackageEntry::to_dict() const {
    Json out = Json::object();
    out["path"] = path;
    out["sha256"] = sha256;
    out["size_bytes"] = size_bytes;
    out["kind"] = kind;
    return out;
}

PackageEntry PackageEntry::from_dict(const Json& data) {
    PackageEntry entry;
    entry.path = py_str(data.at("path"));
    entry.sha256 = py_str(data.at("sha256"));
    entry.size_bytes = py_llong(data.at("size_bytes"));
    entry.kind = data.contains("kind") ? py_str(data.at("kind")) : "artifact";
    return entry;
}

Json PackageManifest::to_dict() const {
    Json out = Json::object();
    out["kind"] = kind;
    out["schema_version"] = schema_version;
    Json project = Json::object();
    project["name"] = project_name;
    project["file"] = project_file;
    out["project"] = project;
    out["created_at"] = created_at;
    Json application = Json::object();
    application["name"] = application_name;
    application["version"] = application_version;
    out["application"] = application;
    Json entry_list = Json::array();
    for (const auto& entry : entries) {
        entry_list.push_back(entry.to_dict());
    }
    out["entries"] = entry_list;
    out["external_dependencies"] = external_dependencies;
    out["missing_dependencies"] = missing_dependencies;
    Json outputs = Json::array();
    for (const auto& output : generated_outputs) {
        outputs.push_back(output);
    }
    out["generated_outputs"] = outputs;
    out["provenance"] = provenance;
    out["options"] = options;
    out["total_size_bytes"] = total_size_bytes;
    out["entry_count"] = static_cast<long long>(entries.size());
    return out;
}

PackageManifest PackageManifest::from_dict(const Json& data) {
    const Json project = data.contains("project") && data.at("project").is_object()
        ? data.at("project")
        : Json::object();
    const Json application =
        data.contains("application") && data.at("application").is_object()
        ? data.at("application")
        : Json::object();

    const Json raw_version = data.contains("schema_version")
        ? data.at("schema_version")
        : Json(0);
    if (raw_version.is_boolean() || !raw_version.is_number_integer()) {
        throw std::invalid_argument("schema_version 必须是整数，得到 "
                                    + repr_scalar(raw_version));
    }

    PackageManifest manifest;
    manifest.kind = data.contains("kind") ? py_str(data.at("kind")) : kManifestKind;
    manifest.schema_version = raw_version.get<int>();
    manifest.project_name = project.contains("name") ? py_str(project.at("name")) : "";
    manifest.project_file = project.contains("file") ? py_str(project.at("file")) : "";
    manifest.created_at = data.contains("created_at") ? py_str(data.at("created_at")) : "";
    manifest.application_name =
        application.contains("name") ? py_str(application.at("name")) : "paleo-workbench";
    manifest.application_version =
        application.contains("version") ? py_str(application.at("version")) : "";
    // Python is fail-closed here: list(5) / dict(5) / int(None) raise
    // TypeError/ValueError, which the verifiers surface as corrupt-manifest.
    // Silently skipping malformed fields could flip a verify verdict.
    if (data.contains("entries")) {
        const Json& entries = data.at("entries");
        if (!entries.is_array()) {
            throw std::invalid_argument("'int' object is not iterable");
        }
        for (const auto& item : entries) {
            manifest.entries.push_back(PackageEntry::from_dict(item));
        }
    }
    if (data.contains("external_dependencies")) {
        if (!data.at("external_dependencies").is_array()) {
            throw std::invalid_argument("'int' object is not iterable");
        }
        manifest.external_dependencies = data.at("external_dependencies");
    }
    if (data.contains("missing_dependencies")) {
        if (!data.at("missing_dependencies").is_array()) {
            throw std::invalid_argument("'int' object is not iterable");
        }
        manifest.missing_dependencies = data.at("missing_dependencies");
    }
    if (data.contains("generated_outputs")) {
        if (!data.at("generated_outputs").is_array()) {
            throw std::invalid_argument("'int' object is not iterable");
        }
        for (const auto& item : data.at("generated_outputs")) {
            manifest.generated_outputs.push_back(py_str(item));
        }
    }
    if (data.contains("provenance")) {
        if (!data.at("provenance").is_object()) {
            throw std::invalid_argument("'" +
                std::string(data.at("provenance").is_null() ? "NoneType" : "int") +
                "' object is not iterable");
        }
        manifest.provenance = data.at("provenance");
    }
    if (data.contains("options")) {
        if (!data.at("options").is_object()) {
            throw std::invalid_argument("argument must be a dict, not " +
                std::string(data.at("options").is_null() ? "NoneType" : "int"));
        }
        manifest.options = data.at("options");
    }
    if (data.contains("total_size_bytes")
        && !data.at("total_size_bytes").is_null()) {
        manifest.total_size_bytes =
            py_llong(data.at("total_size_bytes"));
    }
    return manifest;
}

std::string PackageManifest::dumps() const { return to_dict().dump(1); }

std::set<std::string> PackageManifest::entry_paths() const {
    std::set<std::string> paths;
    for (const auto& entry : entries) {
        paths.insert(entry.path);
    }
    return paths;
}

void PackageManifest::validate_paths() const {
    std::set<std::string> seen_casefold;
    std::set<std::string> seen_nfc;
    for (const auto& entry : entries) {
        const std::string pure =
            safe_relative_path(entry.path, "manifest entry");
        check_collision(pure, seen_casefold, seen_nfc);
    }
    if (!project_file.empty()) {
        safe_relative_path(project_file, "manifest project file");
    }
}

std::filesystem::path write_manifest(const PackageManifest& manifest,
                                     const std::filesystem::path& package_root) {
    manifest.validate_paths();
    const std::filesystem::path target = package_root / kManifestFilename;
    const std::string payload = manifest.dumps();
    const std::filesystem::path temp = package_root
        / (std::string(".") + kManifestFilename + ".tmp");
    {
        std::FILE* handle = std::fopen(temp.string().c_str(), "wb");
        if (handle == nullptr) {
            throw std::runtime_error("cannot write manifest temp: "
                                     + temp.string());
        }
        const std::size_t written = std::fwrite(payload.data(), 1, payload.size(),
                                                handle);
        std::fclose(handle);
        if (written != payload.size()) {
            throw std::runtime_error("short write on manifest temp");
        }
    }
    std::filesystem::rename(temp, target);
    return target;
}

PackageManifest read_manifest(const std::filesystem::path& package_root) {
    const std::filesystem::path path = package_root / kManifestFilename;
    std::FILE* handle = std::fopen(path.string().c_str(), "rb");
    if (handle == nullptr) {
        throw std::runtime_error("cannot open manifest: " + path.string());
    }
    std::string payload;
    char buffer[8192];
    std::size_t chunk = 0;
    while ((chunk = std::fread(buffer, 1, sizeof(buffer), handle)) > 0) {
        payload.append(buffer, chunk);
    }
    std::fclose(handle);
    Json data;
    try {
        data = Json::parse(payload);
    } catch (const Json::parse_error&) {
        throw;  // raw parser error; verifier.py adds the single prefix
    }
    PackageManifest manifest = PackageManifest::from_dict(data);
    manifest.validate_paths();
    return manifest;
}

}  // namespace pwb::interchange

File: legacy/python_reference/product/paleo_workbench/interchange/package/manifest.py
"""Package manifest V2: portable, checksummed, path-safe.

Rules:
- Every path is a package-relative POSIX path, validated by
  :func:`paleo_workbench.interchange.path_safety.safe_relative_path` on both
  write and read. Machine-local absolute paths may appear ONLY in the
  explicit ``external_dependencies`` list (that is their purpose) — never as
  the location of packaged content.
- Content integrity: sha256 + byte size per entry; ``total_size_bytes`` for
  quick capacity checks.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from paleo_workbench.interchange.path_safety import safe_relative_path

MANIFEST_FILENAME = "manifest.json"
MANIFEST_KIND = "paleo-package"
MANIFEST_SCHEMA_VERSION = 2
SUPPORTED_SCHEMA_VERSIONS = (2,)


@dataclass
class PackageEntry:
    path: str  # package-relative POSIX
    sha256: str
    size_bytes: int
    kind: str = "artifact"  # "project" | "artifact" | "metadata"

    def to_dict(self) -> dict[str, Any]:
        return {
            "path": self.path,
            "sha256": self.sha256,
            "size_bytes": self.size_bytes,
            "kind": self.kind,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "PackageEntry":
        return cls(
            path=str(data["path"]),
            sha256=str(data["sha256"]),
            size_bytes=int(data["size_bytes"]),
            kind=str(data.get("kind", "artifact")),
        )


@dataclass
class PackageManifest:
    kind: str = MANIFEST_KIND
    schema_version: int = MANIFEST_SCHEMA_VERSION
    project_name: str = ""
    project_file: str = ""  # package-relative path of the .paleo.json
    created_at: str = ""
    application_name: str = "paleo-workbench"
    application_version: str = ""
    entries: list[PackageEntry] = field(default_factory=list)
    external_dependencies: list[dict[str, Any]] = field(default_factory=list)
    missing_dependencies: list[dict[str, Any]] = field(default_factory=list)
    generated_outputs: list[str] = field(default_factory=list)  # version ids
    provenance: dict[str, Any] = field(default_factory=dict)
    options: dict[str, Any] = field(default_factory=dict)
    total_size_bytes: int = 0

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "schema_version": self.schema_version,
            "project": {"name": self.project_name, "file": self.project_file},
            "created_at": self.created_at,
            "application": {
                "name": self.application_name,
                "version": self.application_version,
            },
            "entries": [entry.to_dict() for entry in self.entries],
            "external_dependencies": list(self.external_dependencies),
            "missing_dependencies": list(self.missing_dependencies),
            "generated_outputs": list(self.generated_outputs),
            "provenance": dict(self.provenance),
            "options": dict(self.options),
            "total_size_bytes": self.total_size_bytes,
            "entry_count": len(self.entries),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "PackageManifest":
        project = data.get("project") or {}
        application = data.get("application") or {}
        raw_version = data.get("schema_version", 0)
        if isinstance(raw_version, bool) or not isinstance(raw_version, int):
            raise ValueError(f"schema_version 必须是整数，得到 {raw_version!r}")
        return cls(
            kind=str(data.get("kind", MANIFEST_KIND)),
            schema_version=raw_version,
            project_name=str(project.get("name", "")),
            project_file=str(project.get("file", "")),
            created_at=str(data.get("created_at", "")),
            application_name=str(application.get("name", "paleo-workbench")),
            application_version=str(application.get("version", "")),
            entries=[PackageEntry.from_dict(e) for e in data.get("entries", ())],
            external_dependencies=list(data.get("external_dependencies", ())),
            missing_dependencies=list(data.get("missing_dependencies", ())),
            generated_outputs=[str(v) for v in data.get("generated_outputs", ())],
            provenance=dict(data.get("provenance", {})),
            options=dict(data.get("options", {})),
            total_size_bytes=int(data.get("total_size_bytes", 0)),
        )

    def dumps(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=1)

    def entry_paths(self) -> set[str]:
        return {entry.path for entry in self.entries}

    def validate_paths(self) -> None:
        """Fail-closed: every stored path must be a safe, collision-free
        relative path (duplicates after NFC/casefold are rejected)."""
        from paleo_workbench.interchange.path_safety import check_collision

        seen_casefold: set[str] = set()
        seen_nfc: set[str] = set()
        for entry in self.entries:
            pure = safe_relative_path(entry.path, what="manifest entry")
            check_collision(pure, seen_casefold, seen_nfc)
        if self.project_file:
            safe_relative_path(self.project_file, what="manifest project file")


def write_manifest(manifest: PackageManifest, package_root: Path) -> Path:
    manifest.validate_paths()
    target = Path(package_root) / MANIFEST_FILENAME
    payload = manifest.dumps()
    from paleo_workbench.resources.exporters import atomic_output

    with atomic_output(target) as tmp:
        tmp.write_text(payload, encoding="utf-8")
    return target


def read_manifest(package_root: Path) -> PackageManifest:
    """Parse manifest.json; raises ValueError on corrupt JSON or unsafe paths."""
    path = Path(package_root) / MANIFEST_FILENAME
    payload = json.loads(path.read_text(encoding="utf-8"))
    manifest = PackageManifest.from_dict(payload)
    manifest.validate_paths()
    return manifest
